### ISSUE TITLE
Setting AdvValidLifetime or any AdvSomething parameters for IPv6 RA config breaks RADVD.

### ISSUE THREAD

**Issue by kafeinnet** [opnsense/core#3589]
**Important notices**
Before you add a new report, we ask you kindly to acknowledge the following:

[-] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md

[-] I have searched the existing issues and I'm convinced that mine is new.

**Describe the bug**
Setting AdvValidLifetime or any AdvSomething parameters for IPv6 RA config breaks RADVD.

Those options are set globally in /var/etc/radvd.conf but they are prefix specific options (man 5 radvd.conf).

This bug is related to the issue #3366.

**To Reproduce**
1. Go to 'Services/Router Advertisements/[any interface]'
2. Click on 'Advanced settings'
3. Set a value to 'AdvValidLifetime'
4. Restart the RA services

**Environment**
OPNsense 19.7-amd64

**Comment by maurice-w** [comment]
Sorry, I hadn't noticed that @AdSchellevis closed #3366 a month ago. As the creator of this feature request I should have tested the implementation, but for some reason I haven't been notified.

It's a bit more complex:
- `AdvDefaultLifetime` is interface specific
- `AdvValidLifetime` and `AdvPreferredLifetime` are prefix specific
- `AdvRDNSSLifetime` is RDNSS specific
- `AdvDNSSLLifetime` is DNSSL specific
- `AdvRouteLifetime` is route specific

It would also be better to have these lifetime fields next to the corresponding option fields in the UI.

**Comment by AdSchellevis** [comment]
@maurice-w the way the code is constructed, there are boundaries in the ways we can change things without risking rebuilding the module as a whole. If you can create some test configs and examples on how they would be different from what is generated now, I'll gladly take a look if we can change it.


With some random numbers in the config, it currently generates a config like:

```
interface XXXX {
        AdvSendAdvert on;
        MinRtrAdvInterval 200;
        MaxRtrAdvInterval 600;
        AdvDefaultLifetime 601;
        AdvValidLifetime 602;
        AdvPreferredLifetime 603;
        AdvRDNSSLifetime 604;
        AdvDNSSLLifetime 605;
        AdvLinkMTU 1500;
        AdvDefaultPreference medium;
        prefix XXXXXXXX {
                DeprecatePrefix on;
                AdvOnLink off;
                AdvAutonomous off;
                AdvRouterAddr on;
        };
        route XXXXXX {
                AdvRouteLifetime 606;
        };
        RDNSS XXXXXX { };
        DNSSL localdomain { };
};
```

Moving settings to the correct paragraph should be easy (like AdvValidLifetime to `prefic {}`), since routes aren't stored very convenient, we can't add per prefix/route options within the scope of this ticket.

If I understand your response correctly, it should have created a config like:

```
interface XXXX {
        AdvSendAdvert on;
        MinRtrAdvInterval 200;
        MaxRtrAdvInterval 600;
        AdvDefaultLifetime 601;
        AdvRDNSSLifetime 604;
        AdvDNSSLLifetime 605;
        AdvLinkMTU 1500;
        AdvDefaultPreference medium;
        prefix XXXXXXXX {
                DeprecatePrefix on;
                AdvOnLink off;
                AdvAutonomous off;
                AdvRouterAddr on;
                AdvValidLifetime 602;
        };
        route XXXXXX {
                AdvRouteLifetime 606;
        };
        RDNSS XXXXXX { };
        DNSSL localdomain { };
};
```

**Comment by AdSchellevis** [comment]
@maurice-w the way the code is constructed, there are boundaries in the ways we can change things without risking rebuilding the module as a whole.

With some random numbers in the config, it currently generates a config like:

```
interface XXXX {
        AdvSendAdvert on;
        MinRtrAdvInterval 200;
        MaxRtrAdvInterval 600;
        AdvDefaultLifetime 601;
        AdvValidLifetime 602;
        AdvPreferredLifetime 603;
        AdvRDNSSLifetime 604;
        AdvDNSSLLifetime 605;
        AdvLinkMTU 1500;
        AdvDefaultPreference medium;
        prefix XXXXXXXX {
                DeprecatePrefix on;
                AdvOnLink off;
                AdvAutonomous off;
                AdvRouterAddr on;
        };
        route XXXXXX {
                AdvRouteLifetime 606;
        };
        RDNSS XXXXXX { };
        DNSSL localdomain { };
};
```

Moving settings to the correct paragraph should be easy (like AdvValidLifetime to `prefic {}`), since routes aren't stored very convenient, we can't add per prefix/route options within the scope of this ticket.

If I understand your response correctly, it should have created a config like:

```
interface XXXX {
        AdvSendAdvert on;
        MinRtrAdvInterval 200;
        MaxRtrAdvInterval 600;
        AdvDefaultLifetime 601;
        AdvLinkMTU 1500;
        AdvDefaultPreference medium;
        prefix XXXXXXXX {
                DeprecatePrefix on;
                AdvOnLink off;
                AdvAutonomous off;
                AdvRouterAddr on;
                AdvValidLifetime 602;
                AdvPreferredLifetime 603;
        };
        route XXXXXX {
                AdvRouteLifetime 606;
        };
        RDNSS XXXXXX {
              AdvRDNSSLifetime 604;
        };
        DNSSL localdomain {
             AdvDNSSLLifetime 605;
        };
};
```


We could discuss  moving some settings around in the UI if that's more logical, although they should stay on the same page (services_router_advertisements.php) to keep it manageable.

**Comment by maurice-w** [comment]
@AdSchellevis, your second config is looking good. Just a heads up that there can be multiple `prefix` and `route` sections for a single interface.

Agreed that we should keep the settings on the same page. May I suggest:
- 'Default Router Lifetime' close to the 'Advertise Default Gateway' checkbox (because unchecking this sets this lifetime to 0 anyway).
- 'Prefix Valid Lifetime' and 'Prefix Preferred Lifetime' wherever (because the prefixes are autogenerated and currently not shown in the UI).
- RDNSS Lifetime close to the 'DNS servers' fields and 'DNSSL Lifetime' close to the 'Domain search list' field (obviously).
- 'Route Lifetime' in the 'Advertise Routes' section (one field for each route if possible, a single field which applies to all routes otherwise).

Also, using friendly names ('Default Router Lifetime' instead of 'AdvDefaultLifetime') would make the UI more consistent. And help texts would be nice (I could write those).

**Comment by AdSchellevis** [comment]
@maurice-w I'll change the backend code to make sure the configuration looks good, I can certainly use help with the UI part. Can you create a pull request for that page?

**Comment by AdSchellevis** [comment]
https://github.com/opnsense/core/commit/d48cfed48cc458359ec940036cd67902a75fb3f1 should fix the functionality.
```
opnsense-patch d48cfed
```

**Comment by maurice-w** [comment]
@AdSchellevis, thanks for fixing the functionality. I will test the patch and report back.
As for updating the website: I took a look at `services_router_advertisements.php` and don't think that's something I can do on short notice. So this might have to be done by someone else or wait until I find the time to familiarize myself with the way the OPNsense UI does things internally (and PHP in general, actually).
Always willing to help with what should go where, writing help texts and so on.

**Comment by AdSchellevis** [comment]
@maurice-w ok, no problem, but if you can dump the proposed changes into a new issue, I can probably change the layout parts for you. 

Something like:

1. (unchanged) "Advertise Default Gateway"
2. [move] 'Default Router Lifetime' 
--> suggested help text if different
3.  .. next item

**Comment by AdSchellevis** [comment]
https://github.com/opnsense/core/commit/2a30428da3f6ac6fc73292941e82d1a7565365a6 shipped in 19.7.1
